### ISSUE TITLE
1.12.0 build fails with system mbedtls 2.11.0

### ISSUE THREAD

**Issue by ilovezfs** [mongrel2/mongrel2#327]
```
iMac-TMP:~ joe$ brew install -dvs mongrel2
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/mongrel2.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/mbedtls.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/zeromq.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cmake.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sphinx-doc.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/asciidoc.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/autoconf.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/docbook-xsl.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/docbook.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/xmlto.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gnu-getopt.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gettext.rb
/usr/bin/sandbox-exec -f /private/tmp/homebrew20180704-34035-1f9bzvf.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew:/usr/local/Homebrew/Library/Homebrew/cask/lib -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/mongrel2.rb --verbose --debug
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/mongrel2.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/mbedtls.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/zeromq.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cmake.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sphinx-doc.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/asciidoc.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/autoconf.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/docbook-xsl.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/docbook.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/xmlto.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gnu-getopt.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gettext.rb
==> Downloading https://github.com/mongrel2/mongrel2/releases/download/v1.12.0/mongrel2-v1.12.0.tar.bz2
Already downloaded: /Users/joe/Library/Caches/Homebrew/mongrel2-1.12.0.tar.bz2
==> Verifying mongrel2-1.12.0.tar.bz2 checksum
tar xjf /Users/joe/Library/Caches/Homebrew/mongrel2-1.12.0.tar.bz2
==> make all
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/adt/darray.o src/adt/darray.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/adt/dict.o src/adt/dict.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/adt/hash.o src/adt/hash.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/adt/list.o src/adt/list.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/adt/radixmap.o src/adt/radixmap.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/adt/tst.o src/adt/tst.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/bstr/bsafe.o src/bstr/bsafe.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/bstr/bstraux.o src/bstr/bstraux.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/bstr/bstrlib.o src/bstr/bstrlib.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/config/config.o src/config/config.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/config/db.o src/config/db.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/config/module.o src/config/module.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC     -c -o src/http11/http11_parser.o src/http11/http11_parser.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/http11/httpclient_parser.o src/http11/httpclient_parser.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/mem/halloc.o src/mem/halloc.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/task/context.o src/task/context.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/task/fd.o src/task/fd.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/task/net.o src/task/net.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/task/qlock.o src/task/qlock.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/task/rendez.o src/task/rendez.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/task/task.o src/task/task.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/bsd_specific.o src/bsd_specific.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/cache.o src/cache.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/chunked.o src/chunked.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/connection.o src/connection.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/control.o src/control.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/dbg.o src/dbg.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/dir.o src/dir.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/filter.o src/filter.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/handler.o src/handler.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/handler_parser.o src/handler_parser.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/headers.o src/headers.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/host.o src/host.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/io.o src/io.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/log.o src/log.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/logrotate.o src/logrotate.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/mime.o src/mime.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/pattern.o src/pattern.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/proxy.o src/proxy.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/register.o src/register.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/request.o src/request.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/response.o src/response.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/routing.o src/routing.c
clang -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread   -fPIC    -c -o src/server.o src/server.c
src/server.c:176:29: error: use of undeclared identifier 'mbedtls_havege_state'; did you mean 'mbedtls_sha512_starts'?
        ctx = calloc(sizeof(mbedtls_havege_state), 1);
                            ^~~~~~~~~~~~~~~~~~~~
                            mbedtls_sha512_starts
/usr/local/include/mbedtls/sha512.h:158:25: note: 'mbedtls_sha512_starts' declared here
MBEDTLS_DEPRECATED void mbedtls_sha512_starts( mbedtls_sha512_context *ctx,
                        ^
src/server.c:177:52: error: expected expression
        mbedtls_havege_init((mbedtls_havege_state *)ctx);
                                                   ^
src/server.c:177:30: error: use of undeclared identifier 'mbedtls_havege_state'; did you mean 'mbedtls_havege_init'?
        mbedtls_havege_init((mbedtls_havege_state *)ctx);
                             ^~~~~~~~~~~~~~~~~~~~
                             mbedtls_havege_init
src/server.c:177:9: note: 'mbedtls_havege_init' declared here
        mbedtls_havege_init((mbedtls_havege_state *)ctx);
        ^
src/server.c:179:25: error: use of undeclared identifier 'mbedtls_havege_random'; did you mean 'mbedtls_havege_init'?
        srv->rng_func = mbedtls_havege_random;
                        ^~~~~~~~~~~~~~~~~~~~~
                        mbedtls_havege_init
src/server.c:177:9: note: 'mbedtls_havege_init' declared here
        mbedtls_havege_init((mbedtls_havege_state *)ctx);
        ^
4 errors generated.
make: *** [src/server.o] Error 1
/usr/local/Homebrew/Library/Homebrew/debrew.rb:11:in `raise'
BuildError: Failed executing: make all
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: 
```

**Comment by ketaro** [comment]
Also on FreeBSD 11.2

**Comment by jasom** [comment]
@ketaro @ilovezfs can you try with the PR in #328?  That should remove the hard requirement on HAVEGE.

**Comment by ketaro** [comment]
@jasom no luck (FreeBSD installs mbedtls-2.9.0).  

```
cc -g -O2 -Wall -Wextra -I./src -DNDEBUG -D_FILE_OFFSET_BITS=64 -pthread -I/usr/local/include -L /usr/local/lib -L/usr/local/lib -pthread src/mongrel2.o -o bin/mongrel2 build/libm2.a -lzmq -lsqlite3 -L /usr/local/lib -L/usr/local/lib -pthread
build/libm2.a(server.o): In function `Server_init_rng':
/root/mongrel2/src/server.c:172: undefined reference to `mbedtls_ctr_drbg_init'
/root/mongrel2/src/server.c:173: undefined reference to `mbedtls_ctr_drbg_seed'
/root/mongrel2/src/server.c:177: undefined reference to `mbedtls_ctr_drbg_random'
build/libm2.a(server.o): In function `Server_init_ssl':
/root/mongrel2/src/server.c:221: undefined reference to `mbedtls_x509_crt_parse_file'
/root/mongrel2/src/server.c:224: undefined reference to `mbedtls_pk_parse_keyfile'
/root/mongrel2/src/server.c:233: undefined reference to `mbedtls_x509_crt_parse_file'
build/libm2.a(server.o): In function `Server_load_ciphers':
/root/mongrel2/src/server.c:123: undefined reference to `mbedtls_ssl_ciphersuite_from_id'
/root/mongrel2/src/server.c:131: undefined reference to `mbedtls_ssl_ciphersuite_from_string'
build/libm2.a(server.o): In function `Server_init_ssl':
/root/mongrel2/src/server.c:246: undefined reference to `mbedtls_ssl_list_ciphersuites'
build/libm2.a(server.o): In function `Server_destroy':
/root/mongrel2/src/server.c:344: undefined reference to `mbedtls_x509_crt_free'
/root/mongrel2/src/server.c:345: undefined reference to `mbedtls_pk_free'
build/libm2.a(connection.o): In function `Connection_fingerprint_from_cert':
/root/mongrel2/src/connection.c:287: undefined reference to `mbedtls_ssl_get_peer_cert'
/root/mongrel2/src/connection.c:299: undefined reference to `mbedtls_sha1_starts'
/root/mongrel2/src/connection.c:300: undefined reference to `mbedtls_sha1_update'
/root/mongrel2/src/connection.c:301: undefined reference to `mbedtls_sha1_finish'
build/libm2.a(connection.o): In function `Connection_create':
/root/mongrel2/src/connection.c:1158: undefined reference to `mbedtls_ssl_conf_own_cert'
/root/mongrel2/src/connection.c:1162: undefined reference to `mbedtls_ssl_conf_ca_chain'
/root/mongrel2/src/connection.c:1168: undefined reference to `mbedtls_ssl_conf_sni'
/root/mongrel2/src/connection.c:1170: undefined reference to `mbedtls_ssl_conf_dh_param'
/root/mongrel2/src/connection.c:1171: undefined reference to `mbedtls_ssl_conf_ciphersuites'
```

**Comment by jasom** [comment]
@ketaro that looks like it's just an error in the freebsd build entry in the makefile; try:

```
LDFLAGS="-lmbedtls -lmbedx509 -lmbedcrypto" make freebsd
```

if that works, I can patch this fairly easily

**Comment by ilovezfs** [comment]
@jasom https://github.com/mongrel2/mongrel2/pull/328 works with the Homebrew formula:
https://gist.github.com/ilovezfs/cef882b73213f0386e76acc9fd924fbb

**Comment by ryandesign** [comment]
> can you try with the PR in #328?

Works for me too in MacPorts. It's been 2 years. Anybody want to merge that?

**Comment by jasom** [comment]
Closed by #349
